### ISSUE TITLE
Default Jersey thread pool should resize based on heuristics (e.g. queue depth)

### ISSUE THREAD

**Issue by batsatt** [oracle/helidon#690]
The `ThreadPoolExecutor` used by default does not resize except under extreme conditions, and the algorithm cannot easily be changed. See #689 for context.

An algorithm that can start off with relatively few threads and adapt to load should be investigated.

**Comment by tomas-langer** [comment]
Please do not do this.
The number of threads can be specified using configuration.
For production environments, the initial thread count should be the final thread count.
For dev environment you can live with a small number.

Variable number of threads is a nightmare to tune and troubleshoot. In most production environments you end up with a fixed thread pool anyway to avoid cold starts (experience from tuning WLS environments).

**Comment by batsatt** [comment]
You misunderstand: configuration will continue to work as is. Users would still be completely free to specify max == min to achieve fixed size behavior. When the configuration _does_ allow for growth, however, the hard-wired algorithm in `ThreadPoolExecutor` can only be affected by controlling the queue capacity, requiring a very small queue for growth to ever occur under reasonable conditions.

I was unaware of the specifics of the hard-wired growth algorithm and I strongly suspect others are as well; this can easily lead to very surprising behavior where growth (above min) 'never' occurs (_exactly_ as happened with our previous default configuration). I want to eliminate this surprise by choosing a different growth strategy by default, while adding the ability to select a strategy via configuration.

The goals here are to: 

1. Provide more control over when the pool should grow above the configured min ("core") size, replacing the hard-wired algorithm with a simple growth strategy interface (effectively a named predicate) that can be selected via configuration.
2. Provide 3 simple growth strategy implementations, e.g. `whenQueueIsFull` (prefer queueing over growth, same as hard-wired strategy), `whenPoolHasRoom` (prefer growth over queuing) and one based on some simple heuristic like queue size or growth rate. 
3. Provide reasonable default behavior with sizing based on cores, a large queue size ( _maybe_ based on memory) and the heuristic growth strategy. 

Note that if a fixed pool is configured, the growth strategy will never be used.

We can debate whether the default pool size should be fixed or allow growth. My strong preference is to set a reasonable minimum but allow for growth to defer the need for tuning.

**Comment by spericas** [comment]
I agree with Bryan that the default behavior is somewhat confusing and surprising, especially the fact that the queue, as I understand it, is of size `MAX_VALUE`. I also agree that we need something better than the default we have now. 

 OTOH, it seems to me that if we allow the queue size to be set externally (via config) and pick a default that is not `MAX_VALUE` we can greatly improve the current situation, without the need for a new algorithm. Obviously, finding a _reasonable_ default is the tricky part, but arguably so is finding the default strategy from the ones mentioned by Bryan above. 

In summary, I'm not against the proposal, but perhaps there's an intermediate step that we can consider.

**Comment by batsatt** [comment]
We do allow setting the queue capacity via config; however, trying to control pool growth solely via this parameter is very challenging... 

First, you have to set it small enough that you will fill the queue or else growth will not occur; choosing such a value is non-trivial as it is highly dependent on specific loads. Once you do fill the queue, new threads will start spinning up, but... the moment you hit the max pool size, subsequent requests will be _rejected_ until threads free up. Under continuous load, when you reach this point the queue is no longer helping to buffer load. 

If this is the only knob we expose to manage growth then I strongly suspect most users will choose to use a fixed pool to avoid unnecessary request rejections (I certainly would); if so, the knob is useless for this purpose.

I'm reasonably confident that we can provide a better out-of-the box experience and am exploring options.

**Comment by tomas-langer** [comment]
I agree that the default "growth" is useless. Nevertheless I am not for giving too many options.
I still think that dynamic number of threads is a very bad idea and would vouch for no-growth under all conditions.
I also agree that using queue for growth is not good, as it slows down the scale-up.
Nevertheless trying it on our own is a dangerous path. In WLS, there is also a "clever" thread pool, that tries to dynamically modify thread number.
Nevertheless when you start "cold" WLS in production under very heavy load, it will not manage to start the threads in time and will just fail.

So if you want to come up with some clever solution, it must be really clever, as otherwise we bring more problems.

**Comment by batsatt** [comment]
If you really believe that growth should not be supported at all, then we should eliminate the current core and max pool size configuration elements and replace them with a single value. It makes no sense at all as things stand currently to have both.

However, I don't think this is the right approach: users should be able to control whether or not to use a fixed size. 

The point of this work is _not_ to try to satisfy demanding production cases! For those cases, users will clearly need to do their own tuning and will most likely choose a fixed pool. 

For development, testing and moderate load cases having a _**default**_ that supports growth avoids violating the principle of least surprise. Adding a simple configuration to control growth rate, again with a reasonable default, is consistent with such a model.

**Comment by batsatt** [comment]
Let's take up the discussion on the PR. If we all conclude that the actual implementation is not worth it, so be it, but I'd like to have all the facts in front of us.
